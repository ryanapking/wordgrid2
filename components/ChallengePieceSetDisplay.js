import React, { Component } from 'react';
import { Text, View, StyleSheet, ScrollView } from 'react-native';
import { connect } from 'react-redux';

import Piece from './Piece';
import DrawPieceSection from './DrawPieceSection';

class ChallengePieceSetDisplay extends Component {
  render() {
    const { consumedSquares, pieceSet } = this.props;
    if (!pieceSet) {
      return null;
    }

    const pieceSetArray = Object.keys(pieceSet).map( (key) => pieceSet[key]);

    return (
      <ScrollView style={styles.fullSize} horizontal={true}>
        <View style={styles.pieceSectionContainer}>
          <DrawPieceSection pieces={pieceSetArray} allowDrag={false}/>
        </View>
      </ScrollView>
    );
  }
}

const styles = StyleSheet.create({
  fullSize: {
    width: '100%',
    height: '100%',
  },
  pieceSectionContainer: {
    height: '100%',
    paddingTop: 10,
    paddingBottom: 10,
    aspectRatio: 13,
    backgroundColor: 'green',
  },
  gamePiece: {
    backgroundColor: 'gray',
    maxWidth: '100%',
    maxHeight: '100%',
    aspectRatio: 1,
  },
});

const mapStateToProps = (state) => {
  const { challenge } = state.challengeData;
  return {
    consumedSquares: challenge.consumedSquares,
    pieceSet: challenge.pieceSet,
  };
};

export default connect(mapStateToProps)(ChallengePieceSetDisplay);