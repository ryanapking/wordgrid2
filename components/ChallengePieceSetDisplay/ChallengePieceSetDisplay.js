import React, { useRef, useEffect, useMemo } from 'react';
import { View, StyleSheet, ScrollView } from 'react-native';
import PropTypes from 'prop-types';

import { squaresArrayType } from "../../proptypes";
import DrawPieceSection from '../DrawPieceSection';

const ChallengePieceSetDisplay = props => {
  const { pieceSet, word, consumedSquares } = props;
  const scrollView = useRef(null);
  const pieceSectionContainer = useRef(null);
  const pieceSetArray = useMemo(() => Object.keys(pieceSet).map( (key) => pieceSet[key]), [pieceSet]);

  const scrollToPiece = (pieceSize) => {
    if (!pieceSectionContainer.current || !scrollView.current || !pieceSet) {
      return;
    }

    pieceSectionContainer.current.measure((x, y, width, height) => {
      // this nonsense gets us to the right piece index
      const scrollIndex = pieceSize - 5;

      // calculate piece size and x value to scroll to
      const pieceWidth = width / 13;
      const scrollToX = (scrollIndex * pieceWidth) + (pieceWidth / 2);

      scrollView.current.scrollTo({x: scrollToX});
    });

  };

  useEffect(() => {
    if (consumedSquares.length >= 4) {
      scrollToPiece(consumedSquares.length <= 16 ? consumedSquares.length : 16);
    }
  });

  if (!pieceSetArray.length || word) return null;

  return (
    <ScrollView
      style={styles.fullSize} horizontal={true} ref={scrollView}>
      <View style={styles.pieceSectionContainer} ref={pieceSectionContainer}>
        <DrawPieceSection pieces={pieceSetArray} allowDrag={false}/>
      </View>
    </ScrollView>
  );
};

ChallengePieceSetDisplay.propTypes = {
  consumedSquares: squaresArrayType.isRequired,
  pieceSet: PropTypes.object.isRequired,
  word: PropTypes.string.isRequired,
};

const styles = StyleSheet.create({
  fullSize: {
    width: '100%',
    height: '100%',
  },
  pieceSectionContainer: {
    height: '100%',
    paddingTop: 10,
    paddingBottom: 10,
    aspectRatio: 13,
  },
  gamePiece: {
    backgroundColor: 'gray',
    maxWidth: '100%',
    maxHeight: '100%',
    aspectRatio: 1,
  },
});

export default ChallengePieceSetDisplay;