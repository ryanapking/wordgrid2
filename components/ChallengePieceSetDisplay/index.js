/* istanbul ignore file */
export { default } from './ChallengePieceSetDisplay';