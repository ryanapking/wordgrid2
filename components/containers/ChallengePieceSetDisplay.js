import React, { Component } from 'react';
import { View, StyleSheet, ScrollView } from 'react-native';
import { connect } from 'react-redux';

import DrawPieceSection from './DrawPieceSection';

class ChallengePieceSetDisplay extends Component {
  componentDidUpdate() {
    const { consumedSquares } = this.props;

    if (consumedSquares.length >= 4) {
      const scrollToPiece = consumedSquares.length <= 16 ? consumedSquares.length : 16;
      this._scrollToPiece(scrollToPiece);
    }
  }

  render() {
    const { pieceSet, word } = this.props;

    if (!pieceSet || word) {
      return null;
    }

    const pieceSetArray = Object.keys(pieceSet).map( (key) => pieceSet[key]);

    return (
      <ScrollView
        style={styles.fullSize} horizontal={true} ref={(scrollView) => this.pieceSet = scrollView}>
        <View style={styles.pieceSectionContainer} ref={(view) => this.pieceSectionContainer = view}>
          <DrawPieceSection pieces={pieceSetArray} allowDrag={false}/>
        </View>
      </ScrollView>
    );
  }

  _scrollToPiece(pieceSize) {
    if (!this.pieceSectionContainer || !this.pieceSet) {
      console.log('guard triggered');
      return;
    }

    this.pieceSectionContainer.measure((x, y, width, height) => {
      // this nonsense gets us to the right piece index
      const scrollIndex = pieceSize - 5;

      // calculate piece size and x value to scroll to
      const pieceWidth = width / 13;
      const scrollToX = (scrollIndex * pieceWidth) + (pieceWidth / 2);

      this.pieceSet.scrollTo({x: scrollToX});
    });

  }
}

const styles = StyleSheet.create({
  fullSize: {
    width: '100%',
    height: '100%',
  },
  pieceSectionContainer: {
    height: '100%',
    paddingTop: 10,
    paddingBottom: 10,
    aspectRatio: 13,
  },
  gamePiece: {
    backgroundColor: 'gray',
    maxWidth: '100%',
    maxHeight: '100%',
    aspectRatio: 1,
  },
});

const mapStateToProps = (state) => {
  const { challenge } = state.challengeData;
  return {
    consumedSquares: challenge.consumedSquares,
    pieceSet: challenge.pieceSet,
    word: challenge.word,
  };
};

export default connect(mapStateToProps)(ChallengePieceSetDisplay);