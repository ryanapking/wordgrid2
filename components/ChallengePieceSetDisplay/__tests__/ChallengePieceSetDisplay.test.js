import React from 'react';
import { render } from 'react-native-testing-library';

import ChallengePieceSetDisplay from "..";

// mock child components
jest.mock('../../DrawPieceSection', () => 'DrawPieceSection');

// mock View component with measure
jest.mock('react-native/Libraries/Components/View/View', () => {
  const mockComponent = require('react-native/jest/mockComponent');
  const measure = (measureCallback) => {
    measureCallback(1, 1, 10, 10, 6, 6);
  };
  return mockComponent('react-native/Libraries/Components/View/View', { measure });
});

describe('ChallengePieceSetDisplay', () => {
  it('renders correctly', () => {
    const { toJSON, unmount } = render(
      <ChallengePieceSetDisplay consumedSquares={[]} pieceSet={testPieceSet} word={''} />
    );
    expect(toJSON()).toMatchSnapshot();
    unmount();
  });

  it('renders null if word is passed to props', () => {
    const { toJSON, unmount } = render(
      <ChallengePieceSetDisplay consumedSquares={testConsumedSquares} pieceSet={testPieceSet} word={'hello'} />
    );
    expect(toJSON()).toBeNull();
    unmount();
  });

  it('repositions the view when a new square is consumed', () => {
    // TODO: Test is causing issues. Need to figure out how to mock findNodeHandle, or how to mock our child components
    const { toJSON, unmount } = render(
      <ChallengePieceSetDisplay consumedSquares={testConsumedSquares} pieceSet={testPieceSet} word={''} />
    );
    expect(toJSON()).not.toBeNull();
    unmount();
  });
});

const testConsumedSquares = [
  {rowIndex: 7,columnIndex: 6, letter: "s"},
  {rowIndex: 8,columnIndex: 7, letter: "h"},
  {rowIndex: 8,columnIndex: 6, letter: "e"},
  {rowIndex: 9,columnIndex: 7, letter: "e"},
  {rowIndex: 9,columnIndex: 6, letter: "n"},
];

const testPieceSet = {
  4: [
    ["","","",""],
    ["","","",""],
    ["t","n","c",""],
    ["l","","",""]
  ],
  5: [
    ["","","","d"],
    ["","","s","w"],
    ["","","s","o"],
    ["","","",""]
  ],
  6: [
    ["","","",""],
    ["","","",""],
    ["t","a","g","i"],
    ["","","h","l"]
  ],
  7: [
    ["","","",""],
    ["o","e","n",""],
    ["l","m","y",""],
    ["","","n",""]
  ],
  8: [
    ["e","i","","o"],
    ["","u","s","i"],
    ["","u","","o"],
    ["","","",""]
  ],
  9: [
    ["","","",""],
    ["","","h","h"],
    ["n","","f","u"],
    ["e","f","i","e"]
  ],
  10: [
    ["","i","i",""],
    ["u","m","",""],
    ["","u","t","h"],
    ["","e","p","t"]
  ],
  11: [
    ["","m","i","a"],
    ["","f","f","c"],
    ["","i","","n"],
    ["t","i","r",""]
  ],
  12: [
    ["","g","d",""],
    ["n","h","",""],
    ["h","m","d","u"],
    ["d","e","d","i"]
  ],
  13: [
    ["","e","n","s"],
    ["","e","i","d"],
    ["","w","i","r"],
    ["n","b","d","e"]
  ],
  14: [
    ["h","","","s"],
    ["c","t","e","l"],
    ["l","f","f","p"],
    ["v","t","t","s"]
  ],
  15: [
    ["e","e","t","y"],
    ["","h","h","h"],
    ["t","d","r","m"],
    ["h","e","l","t"]
  ],
  16: [
    ["a","r","t","v"],
    ["i","h","u","t"],
    ["u","n","r","u"],
    ["d","r","e","d"]
  ]
};