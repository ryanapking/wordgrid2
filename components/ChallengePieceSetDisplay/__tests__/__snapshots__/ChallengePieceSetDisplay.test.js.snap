// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChallengePieceSetDisplay renders correctly 1`] = `
<RCTScrollView
  horizontal={true}
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
>
  <View>
    <View
      style={
        Object {
          "aspectRatio": 13,
          "height": "100%",
          "paddingBottom": 10,
          "paddingTop": 10,
        }
      }
      testID="pieceSectionContainer"
    >
      <DrawPieceSection
        allowDrag={false}
        pieces={
          Array [
            Array [
              Array [
                "",
                "",
                "",
                "",
              ],
              Array [
                "",
                "",
                "",
                "",
              ],
              Array [
                "t",
                "n",
                "c",
                "",
              ],
              Array [
                "l",
                "",
                "",
                "",
              ],
            ],
            Array [
              Array [
                "",
                "",
                "",
                "d",
              ],
              Array [
                "",
                "",
                "s",
                "w",
              ],
              Array [
                "",
                "",
                "s",
                "o",
              ],
              Array [
                "",
                "",
                "",
                "",
              ],
            ],
            Array [
              Array [
                "",
                "",
                "",
                "",
              ],
              Array [
                "",
                "",
                "",
                "",
              ],
              Array [
                "t",
                "a",
                "g",
                "i",
              ],
              Array [
                "",
                "",
                "h",
                "l",
              ],
            ],
            Array [
              Array [
                "",
                "",
                "",
                "",
              ],
              Array [
                "o",
                "e",
                "n",
                "",
              ],
              Array [
                "l",
                "m",
                "y",
                "",
              ],
              Array [
                "",
                "",
                "n",
                "",
              ],
            ],
            Array [
              Array [
                "e",
                "i",
                "",
                "o",
              ],
              Array [
                "",
                "u",
                "s",
                "i",
              ],
              Array [
                "",
                "u",
                "",
                "o",
              ],
              Array [
                "",
                "",
                "",
                "",
              ],
            ],
            Array [
              Array [
                "",
                "",
                "",
                "",
              ],
              Array [
                "",
                "",
                "h",
                "h",
              ],
              Array [
                "n",
                "",
                "f",
                "u",
              ],
              Array [
                "e",
                "f",
                "i",
                "e",
              ],
            ],
            Array [
              Array [
                "",
                "i",
                "i",
                "",
              ],
              Array [
                "u",
                "m",
                "",
                "",
              ],
              Array [
                "",
                "u",
                "t",
                "h",
              ],
              Array [
                "",
                "e",
                "p",
                "t",
              ],
            ],
            Array [
              Array [
                "",
                "m",
                "i",
                "a",
              ],
              Array [
                "",
                "f",
                "f",
                "c",
              ],
              Array [
                "",
                "i",
                "",
                "n",
              ],
              Array [
                "t",
                "i",
                "r",
                "",
              ],
            ],
            Array [
              Array [
                "",
                "g",
                "d",
                "",
              ],
              Array [
                "n",
                "h",
                "",
                "",
              ],
              Array [
                "h",
                "m",
                "d",
                "u",
              ],
              Array [
                "d",
                "e",
                "d",
                "i",
              ],
            ],
            Array [
              Array [
                "",
                "e",
                "n",
                "s",
              ],
              Array [
                "",
                "e",
                "i",
                "d",
              ],
              Array [
                "",
                "w",
                "i",
                "r",
              ],
              Array [
                "n",
                "b",
                "d",
                "e",
              ],
            ],
            Array [
              Array [
                "h",
                "",
                "",
                "s",
              ],
              Array [
                "c",
                "t",
                "e",
                "l",
              ],
              Array [
                "l",
                "f",
                "f",
                "p",
              ],
              Array [
                "v",
                "t",
                "t",
                "s",
              ],
            ],
            Array [
              Array [
                "e",
                "e",
                "t",
                "y",
              ],
              Array [
                "",
                "h",
                "h",
                "h",
              ],
              Array [
                "t",
                "d",
                "r",
                "m",
              ],
              Array [
                "h",
                "e",
                "l",
                "t",
              ],
            ],
            Array [
              Array [
                "a",
                "r",
                "t",
                "v",
              ],
              Array [
                "i",
                "h",
                "u",
                "t",
              ],
              Array [
                "u",
                "n",
                "r",
                "u",
              ],
              Array [
                "d",
                "r",
                "e",
                "d",
              ],
            ],
          ]
        }
      />
    </View>
  </View>
</RCTScrollView>
`;
